openmdao/core/system.py

```diff
@@ -3124,57 +3161,68 @@ def _write_outputs(self, in_or_out, comp_type, outputs, hierarchical, print_arra
+                    if name not in var_dict:  # If not in the merged dict, add it
+                        var_dict[name] = proc_vars[name]
+                    else:
+                        # In there already, only need to deal with it if it is a distributed array
+                        # Checking to see if distributed depends on if it is an input or output
+                        if var_type == 'input':
                             is_distributed = meta[name]['src_indices'] is not None
                         else:
                             is_distributed = meta[name]['distributed']
                         if is_distributed:
                             # TODO no support for > 1D arrays
                             #   meta.src_indices has the info we need to piece together arrays
-                            if 'value' in dict_of_outputs[name]:
-                                dict_of_outputs[name]['value'] = \
-                                    np.append(dict_of_outputs[name]['value'],
-                                              proc_outputs[name]['value'])
-                            if 'shape' in dict_of_outputs[name]:
+                            if 'value' in var_dict[name]:
+                                var_dict[name]['value'] = \
+                                    np.append(var_dict[name]['value'],
+                                              proc_vars[name]['value'])
+                            if 'shape' in var_dict[name]:
                                 # TODO might want to use allprocs_abs2meta_out[name]['global_shape']
-                                dict_of_outputs[name]['shape'] = \
-                                    dict_of_outputs[name]['value'].shape
-                            if 'resids' in dict_of_outputs[name]:
-                                dict_of_outputs[name]['resids'] = \
-                                    np.append(dict_of_outputs[name]['resids'],
-                                              proc_outputs[name]['resids'])
-
-        write_outputs(in_or_out, comp_type, dict_of_outputs, hierarchical, print_arrays, out_stream,
-                      self.pathname, self._var_allprocs_abs_names)
+                                var_dict[name]['shape'] = \
+                                    var_dict[name]['value'].shape
+                            if 'resids' in var_dict[name]:
+                                var_dict[name]['resids'] = \
+                                    np.append(var_dict[name]['resids'],
+                                              proc_vars[name]['resids'])
+
+        # get list of var names in execution order, based on the order subsystems were setup
+        var_list = []
+        in_or_out = 'input' if var_type is 'input' else 'output'
+        for subsys in self._subsystems_allprocs:
+            for n in self._var_allprocs_abs_names[in_or_out]:
+                if n in var_dict and n.startswith(subsys.pathname):
```

Comment:
You're looping over _subsystems_allprocs here, but subsys.pathname will only be defined properly if a subsystem is local.  subsys.name *will* be properly defined though, so maybe  if n startswith '.'.join((self.pathname, subsys.name)) + '.'  ?

Reply:
done.  needs a test.